source/solvers/post_processors_smoothing.cc

```diff
@@ -0,0 +1,323 @@
+                        this->locally_owned_dofs,
+                        dsp,
+                        this->mpi_communicator);
+
+  QGauss<dim> quadrature_formula(number_quadrature_points);
+
+  FEValues<dim> fe_values(*this->mapping,
+                          fe_q,
+                          quadrature_formula,
+                          update_values | update_quadrature_points |
+                            update_JxW_values);
+
+  const unsigned int dofs_per_cell = fe_q.dofs_per_cell;
+  const unsigned int n_q_points    = quadrature_formula.size();
+
+  FullMatrix<double> local_matrix(dofs_per_cell, dofs_per_cell);
+  std::vector<types::global_dof_index> local_dof_indices(dofs_per_cell);
+  std::vector<double>                  phi_vf(dofs_per_cell);
+
+  (*system_matrix) = 0;
+
+  for (const auto &cell : dof_handler.active_cell_iterators())
+    {
+      if (cell->is_locally_owned())
+        {
+          fe_values.reinit(cell);
+
+          local_matrix = 0;
+
+          for (unsigned int q = 0; q < n_q_points; ++q)
+            {
+              for (unsigned int k = 0; k < dofs_per_cell; ++k)
+                {
+                  phi_vf[k] = fe_values.shape_value(k, q);
+                }
+              for (unsigned int i = 0; i < dofs_per_cell; ++i)
+                {
+                  // Assemble L2 projection
+                  // Matrix assembly
+                  for (unsigned int j = 0; j < dofs_per_cell; ++j)
+                    {
+                      local_matrix(i, j) +=
+                        (phi_vf[j] * phi_vf[i]) * fe_values.JxW(q);
+                    }
+                }
+            }
+          cell->get_dof_indices(local_dof_indices);
+          constraints.distribute_local_to_global(local_matrix,
+                                                 local_dof_indices,
+                                                 *system_matrix);
+        }
+    }
+  system_matrix->compress(VectorOperation::add);
+}
+
+template <int dim, typename VectorType>
+const TrilinosWrappers::MPI::Vector &
+PostProcessorSmoothing<dim, VectorType>::solve_L2_projection()
+{
+  // Solve the L2 projection system
+  const double linear_solver_tolerance = 1e-15;
+
+  completely_distributed_solution =
+    TrilinosWrappers::MPI::Vector(this->locally_owned_dofs,
+                                  this->mpi_communicator);
+
+  SolverControl solver_control(
+    this->simulation_parameters.linear_solver.max_iterations,
+    linear_solver_tolerance,
+    true,
+    true);
+
+  TrilinosWrappers::SolverCG solver(solver_control);
+
+  //**********************************************
+  // Trillinos Wrapper ILU Preconditioner
+  //*********************************************
+  const double ilu_fill =
+    this->simulation_parameters.linear_solver.ilu_precond_fill;
+  const double ilu_atol =
+    this->simulation_parameters.linear_solver.ilu_precond_atol;
+  const double ilu_rtol =
+    this->simulation_parameters.linear_solver.ilu_precond_rtol;
```

Comment:
I am not sure I would take these parameters from the simulation parameters.
The L2 projection is a trivial matrix, so I would just put default values here and be done with it.
Fill level of 0, atol of 1e-15, rtol of 1 for example would be just fine.